src/main/java/net/minecraftforge/common/world/RegistryAccessExtension.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.world;
+
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.Lifecycle;
+import net.minecraft.core.Registry;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraftforge.registries.DeferredRegister;
+import net.minecraftforge.registries.ForgeRegistryEntry;
+import net.minecraftforge.registries.IForgeRegistryEntry;
+import net.minecraftforge.registries.RegistryBuilder;
+
+/**
+ * Defines an additional world-gen component to load as part of the RegistryAccess registries.
+ * <p>
+ * The path of the provided ResourceKey becomes the data folder from which datapack definitions are parsed
+ * and the provided Codec handles encoding and decoding to and from the data format.
+ * <p>
+ * To avoid conflicts with other mods, the path of the ResourceKey must be prefixed with the owning mod's
+ * namespace (ie {@code <modid>:<modid>/path}). Static helper methods for creating valid ResourceKeys are
+ * provided at {@link #createRegistryKey(String, String)} and {@link #createRegistryKey(ResourceLocation)}.
+ * <p>
+ * Default or 'built-in' registry entries can be registered via a forge DeferredRegister or IForgeRegistry
+ * of the same registry ResourceKey. These elements will populate the resulting RegistryAccess before datapacks
+ * are loaded and may be overridden by data-definitions. A static helper method for creating a DeferredRegisters
+ * for a RegistryAccessExtension is provided at {@link #createRegister(Class, ResourceKey)}.
+ *
+ * @param <T> The type of world-gen component.
+ */
+public class RegistryAccessExtension<T extends IForgeRegistryEntry<T>> extends ForgeRegistryEntry<RegistryAccessExtension<?>>
```

Comment:
Shouldn't this be final? I don't believe there should be a need to expand behavior here.

Reply:
Yep good point. Would be possible to override the ResourceKey to get around the restrictions on the path prefix